### PR TITLE
UF: adding the SK problem

### PR DESCRIPTION
**Pull request by alejomonbar**
[Unitary Fund](http://unitary.fund)

Adding the Sherrington–Kirkpatrick (SK) problem class with its corresponding test.
https://en.wikipedia.org/wiki/Spin_glass

## Description

This pull request adds support for the SK problem, an optimization problem related to spin glass. The SK problem involves finding the minimum energy Hamiltonian based on the interaction between spins.

The implementation includes a new class `SK` in the problem module. The `SK` class allows the creation of instances of the SK problem. It provides methods for solving the problem using the Docplex solver, visualizing the graph with a given solution, and obtaining the qubo.

Example:
```python
from openqaoa.problems import SK
import matplotlib.pyplot as plt
from openqaoa import QAOA

sk = SK.random_instance(n_nodes=5)
sol = sk.classical_solution()

qaoa = QAOA()
qaoa.compile(sk.qubo)
qaoa.optimize()

sol_qaoa = qaoa.result.lowest_cost_bitstrings()["solutions_bitstrings"][0]

fig, ax = plt.subplots(1, 2, figsize = (10, 5))
sk.plot_solution(sol_qaoa, ax=ax[0])
ax[0].set_title("QAOA Results")
sk.plot_solution(sol, ax=ax[1])
ax[1].set_title("CPLEX Results")
```

Screenshot 2023-08-16 at 3 00 18 PM

## Checklist

[//]: <> (- [ ] My code follows the style guidelines of this project)

- [x] I have performed a self-review of my code.
- [x] I have commented my code and used numpy-style docstrings
- [ ] I have made corresponding updates to the documentation.
- [x] My changes generate no new warnings
- [x] I have added/updated tests to make sure bugfix/feature works.
- [x] New and existing unit tests pass locally with my changes.

[//]: <> (- [ ] Any dependent changes have been merged and published in downstream modules)

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

the test is inside './tests/test_sk.py' and './tests/test_qubo.py'

### CODE REVIEW

**Review by shahidee44 — APPROVED** [review]

**Review by shahidee44 — APPROVED** [review]

File: src/openqaoa-core/openqaoa/problems/sherrington_kirkpatrick.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+        model.solve()
+        status = model.solve_details.status
+        if status != 'integer optimal solution':
+            print(status) 
+        if string:
+            return "".join(str(round(model.solution.get_value(var))) for var in model.iter_binary_vars())
+        else:
+            return {var.name: model.solution.get_value(var) for var in model.iter_binary_vars()}
+
+    def plot_solution(self, solution, ax=None):
```
Comment (shahidee44) [inline src/openqaoa-core/openqaoa/problems/sherrington_kirkpatrick.py:133]: This version of the `plot_solution` function outputs 2 graphs when used in jupyterlab.

File: src/openqaoa-core/openqaoa/problems/sherrington_kirkpatrick.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+        """
```
Comment (shahidee44) [inline src/openqaoa-core/openqaoa/problems/sherrington_kirkpatrick.py:134]: I get the following error when doing this:
```
prob.plot_solution(q.result.lowest_cost_bitstrings()['solutions_bitstrings'][0])
```

image

**Review by shahidee44 — CHANGES_REQUESTED** [review]

File: src/openqaoa-core/openqaoa/problems/sherrington_kirkpatrick.py (hunk 1 above)
Comment (alejomonbar) [reply src/openqaoa-core/openqaoa/problems/sherrington_kirkpatrick.py]: done @shahidee44

File: src/openqaoa-core/openqaoa/problems/sherrington_kirkpatrick.py (hunk 2 above)
Comment (alejomonbar) [reply src/openqaoa-core/openqaoa/problems/sherrington_kirkpatrick.py:134]: fixed @shahidee44

**Review by shahidee44 — APPROVED** [review]

**Review by vishal-ph — APPROVED** [review]

**Pull request merged**